frontend_csv/src/scanner/mod.rs

```diff
@@ -1,14 +1,163 @@
+                token::CSVTokenType::ChannelId => assert_eq!(c.val, "channel_id"),
+                token::CSVTokenType::Signature => assert_eq!(c.val, "signature"),
+                token::CSVTokenType::Point => assert_eq!(c.val, "point"),
+                token::CSVTokenType::ChainHash => assert_eq!(c.val, "chain_hash"),
+                token::CSVTokenType::Byte => assert_eq!(c.val, "byte"),
+                token::CSVTokenType::BigSize => assert_eq!(c.val, "bigsize"),
+                token::CSVTokenType::LiteralString => continue,
+                token::CSVTokenType::Number => continue,
+                token::CSVTokenType::Tu32 => assert_eq!(c.val, "tu32"),
+                token::CSVTokenType::Tu64 => assert_eq!(c.val, "tu64"),
+            }
+        }
+    }
+    #[test]
+    fn test_one_line() {
+        let contents = "msgtype,init,16\n";
+        let char_vec: Vec<char> = contents.chars().collect();
+        let mut scanner = scanner::Scanner::new(1);
+        let result = scanner.scan(&char_vec);
+        assert!(result.len() > 0);
+        let expected = vec![
+            token::CSVToken {
+                ty: token::CSVTokenType::MsgTy,
+                val: "msgtype".to_string(),
+            },
+            token::CSVToken {
+                ty: token::CSVTokenType::LiteralString,
+                val: "init".to_string(),
+            },
+            token::CSVToken {
+                ty: token::CSVTokenType::Number,
+                val: "16".to_string(),
+            },
+        ];
+        for c in 0..expected.len() - 1 {
+            assert!(result[c].val == expected[c].val);
+            debug_assert_eq!(result[c].ty, expected[c].ty);
+        }
+    }
+    #[test]
+    fn test_middle_line() {
+        let contents = "msgtype,init,16\nmsgdata,init,gflen,u16,\n
+        msgdata,init,globalfeatures,byte,gflen\n";
+        let char_vec: Vec<char> = contents.chars().collect();
+        let mut scanner = scanner::Scanner::new(1);
+        let result = scanner.scan(&char_vec);
+        assert!(result.len() > 0);
+        let expected = vec![
+            token::CSVToken {
+                ty: token::CSVTokenType::MsgData,
+                val: "msgdata".to_string(),
+            },
+            token::CSVToken {
+                ty: token::CSVTokenType::LiteralString,
+                val: "init".to_string(),
+            },
+            token::CSVToken {
+                ty: token::CSVTokenType::LiteralString,
+                val: "gflen".to_string(),
+            },
+            token::CSVToken {
+                ty: token::CSVTokenType::U16,
+                val: "u16".to_string(),
+            },
+        ];
+        for c in 0..expected.len() - 1 {
+            debug_assert_eq!(result[c + 3].val, expected[c].val);
+            debug_assert_eq!(result[c + 3].ty, expected[c].ty);
+        }
+    }
+    #[test]
+    fn test_last_line() {
+        let contents = "msgtype,init,16\nmsgdata,init,gflen,u16,\n
```

Comment:
```suggestion
        let contents = "msgtype,init,16\nmsgdata,init,gflen,u16,\n
```

You can use the raw string https://stackoverflow.com/questions/26611664/what-is-the-r-operator-in-rust

In addition, you need to document your code, otherwise, I will put macros in the code that make a compilation error if you push some code without comment.

Up to you! Each time I need to spend 30 minutes to read the code, and if you comment will be more easy to me